File: packages/faro-ios-sdk/Sources/FaroSDK/Session/SessionManager.swift
import Foundation

public struct SessionConfig {
    public let enabled: Bool
    public let persistent: Bool
    public let maxSessionDurationSeconds: TimeInterval
    public let sessionTimeoutSeconds: TimeInterval
    public let samplingRate: Double

    public init(
        enabled: Bool = true,
        persistent: Bool = true,
        maxSessionDurationSeconds: TimeInterval = 4 * 60 * 60,
        sessionTimeoutSeconds: TimeInterval = 15 * 60,
        samplingRate: Double = 1.0
    ) {
        self.enabled = enabled
        self.persistent = persistent
        self.maxSessionDurationSeconds = maxSessionDurationSeconds
        self.sessionTimeoutSeconds = sessionTimeoutSeconds
        self.samplingRate = samplingRate
    }
}

internal final class SessionManager {
    private let config: SessionConfig
    private let store: SessionStore
    private let logger: InternalLogger
    private let lock = NSLock()

    private var currentSessionId: String?
    private var sessionStartTime: Date = .distantPast
    private var lastActivityTime: Date = .distantPast
    private var _isSampled: Bool = true

    var isSampled: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isSampled
    }

    init(config: SessionConfig, store: SessionStore, logger: InternalLogger) {
        self.config = config
        self.store = store
        self.logger = logger
    }

    func start() -> String {
        lock.lock()
        defer { lock.unlock() }
        return lockedStart()
    }

    private func lockedStart() -> String {
        guard config.enabled else {
            logger.debug("Session tracking disabled")
            return ""
        }

        if config.persistent, let stored = store.loadSession(), !isExpired(stored) {
            currentSessionId = stored.sessionId
            sessionStartTime = stored.startTime
            lastActivityTime = Date()
            _isSampled = stored.isSampled
            logger.debug("Restored session: \(currentSessionId ?? "")")
            return currentSessionId!
        }

        return lockedCreateNewSession()
    }

    func getSessionId() -> String {
        lock.lock()
        defer { lock.unlock() }

        if currentSessionId == nil {
            return lockedStart()
        }

        let now = Date()
        if now.timeIntervalSince(sessionStartTime) > config.maxSessionDurationSeconds ||
           now.timeIntervalSince(lastActivityTime) > config.sessionTimeoutSeconds {
            return lockedCreateNewSession()
        }

        lastActivityTime = now
        if config.persistent {
            store.saveSession(StoredSession(
                sessionId: currentSessionId!,
                startTime: sessionStartTime,
                isSampled: _isSampled
            ))
        }

        return currentSessionId!
    }

    func setSessionId(_ sessionId: String) {
        lock.lock()
        defer { lock.unlock() }

        currentSessionId = sessionId
        sessionStartTime = Date()
        lastActivityTime = sessionStartTime
        _isSampled = true
        if config.persistent {
            store.saveSession(StoredSession(sessionId: sessionId, startTime: sessionStartTime, isSampled: true))
        }
    }

    private func lockedCreateNewSession() -> String {
        let sessionId = generateSessionId()
        currentSessionId = sessionId
        sessionStartTime = Date()
        lastActivityTime = sessionStartTime
        _isSampled = Double.random(in: 0...1) < config.samplingRate

        if config.persistent {
            store.saveSession(StoredSession(sessionId: sessionId, startTime: sessionStartTime, isSampled: _isSampled))
        }

        logger.debug("Created new session: \(sessionId) (sampled: \(_isSampled))")
        return sessionId
    }

    private func isExpired(_ stored: StoredSession) -> Bool {
        return Date().timeIntervalSince(stored.startTime) > config.maxSessionDurationSeconds
    }

    private func generateSessionId() -> String {
        return UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(16).lowercased()
    }
}

File: packages/faro-android-sdk/src/main/kotlin/com/grafana/faro/session/SessionManager.kt
package com.grafana.faro.session

import com.grafana.faro.internal.InternalLogger
import java.util.UUID

data class SessionConfig(
    val enabled: Boolean = true,
    val persistent: Boolean = true,
    val maxSessionDurationMs: Long = 4 * 60 * 60 * 1000, // 4 hours
    val sessionTimeoutMs: Long = 15 * 60 * 1000,          // 15 minutes inactivity
    val samplingRate: Double = 1.0
)

internal class SessionManager(
    private val config: SessionConfig,
    private val store: SessionStore,
    private val logger: InternalLogger
) {
    private var currentSessionId: String? = null
    private var sessionStartTime: Long = 0
    private var lastActivityTime: Long = 0
    private var isSampled: Boolean = true

    fun start(): String {
        if (!config.enabled) {
            logger.debug("Session tracking disabled")
            return ""
        }

        // Check for persisted session
        if (config.persistent) {
            val stored = store.loadSession()
            if (stored != null && !isExpired(stored)) {
                currentSessionId = stored.sessionId
                sessionStartTime = stored.startTime
                lastActivityTime = System.currentTimeMillis()
                isSampled = stored.isSampled
                logger.debug("Restored session: $currentSessionId")
                return currentSessionId!!
            }
        }

        return createNewSession()
    }

    fun getSessionId(): String {
        if (currentSessionId == null) {
            return start()
        }

        // Check if session needs rotation
        val now = System.currentTimeMillis()
        if (now - sessionStartTime > config.maxSessionDurationMs ||
            now - lastActivityTime > config.sessionTimeoutMs
        ) {
            return createNewSession()
        }

        lastActivityTime = now
        if (config.persistent) {
            store.saveSession(StoredSession(currentSessionId!!, sessionStartTime, isSampled))
        }

        return currentSessionId!!
    }

    fun isSampled(): Boolean = isSampled

    fun setSessionId(sessionId: String) {
        currentSessionId = sessionId
        sessionStartTime = System.currentTimeMillis()
        lastActivityTime = sessionStartTime
        isSampled = true
        if (config.persistent) {
            store.saveSession(StoredSession(sessionId, sessionStartTime, isSampled))
        }
    }

    private fun createNewSession(): String {
        currentSessionId = generateSessionId()
        sessionStartTime = System.currentTimeMillis()
        lastActivityTime = sessionStartTime
        isSampled = Math.random() < config.samplingRate

        if (config.persistent) {
            store.saveSession(StoredSession(currentSessionId!!, sessionStartTime, isSampled))
        }

        logger.debug("Created new session: $currentSessionId (sampled: $isSampled)")
        return currentSessionId!!
    }

    private fun isExpired(stored: StoredSession): Boolean {
        val now = System.currentTimeMillis()
        return now - stored.startTime > config.maxSessionDurationMs
    }

    private fun generateSessionId(): String {
        return UUID.randomUUID().toString().replace("-", "").take(16)
    }
}
